### PR TITLE
451006-MaksimBarouski

### PR DESCRIPTION
**Pull request by maximka777**

### CODE REVIEW

**Comment by maximka777** [comment]
Т. к. я сразу выполнял лабораторную в своём репозитории, то весь процесс работы здесь https://github.com/maximka777/MPP-Tracer-Test

File: Tracer/Tracer/ConsoleTraceResultFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Tracer
+{
+    public class ConsoleTraceResultFormatter : ITraceResultFormatter
+    {
+        private static string GetTabSequence(int count)
+        {
+            string result = "";
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/ConsoleTraceResultFormatter.cs]: @maximka777, consider the usage of `StringBuilder` as you concatenate strings in a loop.

File: Tracer/Tracer/MyTree.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            get 
+            { 
+                return root; 
+            } 
+            set {
+                root = value; 
+            }
+        }
+
+        private IEnumerable<TreeNode> climbNode(TreeNode node)
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/MyTree.cs]: @maximka777 `PascalCase` for method names.

File: Tracer/Tracer/Tracer.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace Tracer
+{
+    public class Tracer : ITracer
+    {
+        private static volatile Tracer instance = null;
+        private static readonly Object syncObj = new Object();
+        private TraceResult traceResult;
+
+        public Tracer()
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/Tracer.cs]: @maximka777 Singleton with public constructor? :)

File: Tracer/Tracer/Tracer.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            traceResult = new TraceResult();
+        }
+
+        public static Tracer Instance()
+        {
+            if (instance == null)
+            {
+                lock (syncObj)
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/Tracer.cs]: @maximka777 take a look at [some other singleton implementations](http://csharpindepth.com/Articles/General/Singleton.aspx)

File: Tracer/Tracer/TreeNode.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Tracer
+{
+    public class TreeNode
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/TreeNode.cs]: @maximka777 though this is a valid implementation for some sort of DTO class a more commonly encountered solution is the usage of auto-implemented properties.

File: Tracer/Tracer/TraceResult.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Reflection;
+
+namespace Tracer
+{
+    public class TraceResult
+    {
+        public ConcurrentDictionary<int, MyTree> threadDictionary = new ConcurrentDictionary<int, MyTree>();
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/TraceResult.cs]: @maximka777, replace with property.

File: Tracer/Tracer/XmlTraceResultFormatter.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+                foreach (TreeNode node in traceResult.threadDictionary[key].climbTree()){
+                    XElement methodElement = new XElement("mathod");
+                    methodElement.SetAttributeValue("methodName", node.methodName);
+                    methodElement.SetAttributeValue("paramsCount", node.paramsCount);
+                    methodElement.SetAttributeValue("totalTime", node.totalTime);
+                    methodElement.SetAttributeValue("className", node.className);
+                    threadElement.Add(methodElement);
+                }
+            }
+            root.Save("xml.xml");
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/XmlTraceResultFormatter.cs]: @maximka777 replace hardcoded string with instance field and make the class accept the path value as a constructor parameter

**Comment by RamanBut-Husaim** [comment]
@maximka777 try to pay more attention to the encapsulation in your code so that many classes and constructors could be `internal` vs `public`.

**Comment by RamanBut-Husaim** [comment]
@maximka777 Have you made all the fixes?

**Comment by maximka777** [comment]
Yes

File: Tracer/Tracer/ConsoleTraceResultFormatter.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Tracer
+{
+    public class ConsoleTraceResultFormatter : ITraceResultFormatter
+    {
+        private static string GetTabSequence(int count)
+        {
+            StringBuilder sb = new StringBuilder("");
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/ConsoleTraceResultFormatter.cs]: @maximka777 Use `string.Empty`.

Comment (RamanBut-Husaim) [inline Tracer/Tracer/ConsoleTraceResultFormatter.cs]: @maximka777 In general, you could omit this completely as it will be empty by default.

Comment (maximka777) [inline Tracer/Tracer/ConsoleTraceResultFormatter.cs]: Ok

**Comment by maximka777** [comment]
Why do you not close my PR?

**Comment by RamanBut-Husaim** [comment]
@maximka777 Have you pushed all the modifications that was made on the lesson?

**Comment by maximka777** [comment]
Yes, I pushed... few minutes ago)

**Comment by RamanBut-Husaim** [comment]
@maximka777 Got it.

**Pull request merged**